### PR TITLE
Master stat buttons generic mka

### PR DESCRIPTION
**Pull request by mka-odoo**
Task - https://www.odoo.com/web?#id=19565&view_type=form&model=project.task&menu_id=3942&action=327
Task Pad - https://pad.odoo.com/p/r.pV9Sp1Daa0Hzb9dZ

### CODE REVIEW

File: addons/crm_partner_assign/res_partner_view.xml [hunk 1]
```diff
@@ -128,7 +128,17 @@
...
+                            <button name="%(crm.relate_partner_opportunities)d"
+                                    type="action" icon="fa-star"
+                                    class="oe_stat_button">
+                                    <div class="o_form_field o_stat_info mr4">
+                                        <span class="o_stat_value"><field name="assinged_opp_count" /></span>
+                                    </div>
+                                    <div class="o_form_field o_stat_info">
+                                        <span class="o_stat_text">Assinged </span>
+                                        <span class="o_stat_text">Opportunities</span>
+                                    </div>
```
Comment (atp-odoo) [inline addons/crm_partner_assign/res_partner_view.xml]: indentation problem.move left

File: addons/hr_equipment/models/hr_equipment.py [hunk 2]
```diff
@@ -126,10 +126,13 @@ def name_search(self, name, args=None, operator='ilike', limit=100):
     maintenance_ids = fields.One2many('hr.equipment.request', 'equipment_id', domain=[('stage_id.fold', '=', False)])
     maintenance_count = fields.Integer(compute='_compute_maintenance_count', string="Maintenance")
 
-    @api.one
+    @api.multi
     @api.depends('maintenance_ids')
```
Comment (atp-odoo) [inline addons/hr_equipment/models/hr_equipment.py]: no need to use @api.depends()

File: addons/hr_holidays/hr_holidays.py [hunk 3]
```diff
@@ -568,14 +568,18 @@ def _get_leave_status(self, cr, uid, ids, name, args, context=None):
...
         date_begin = date.today().replace(day=1)
         date_end = date_begin.replace(day=calendar.monthrange(date_begin.year, date_begin.month)[1])
-        for employee_id in ids:
-            leaves = Holidays.search_count(cr, uid, [('employee_id', '=', employee_id), ('type', '=', 'remove')], context=context)
-            approved_leaves = Holidays.search_count(cr, uid, [('employee_id', '=', employee_id), ('type', '=', 'remove'), ('date_from', '>=', date_begin.strftime(tools.DEFAULT_SERVER_DATE_FORMAT)), ('date_from', '<=', date_end.strftime(tools.DEFAULT_SERVER_DATE_FORMAT)), ('state', '=', 'validate'), ('payslip_status', '=', False)], context=context)
-            res[employee_id] = {'leaves_count': leaves, 'approved_leaves_count': approved_leaves}
+        approved_leaves_data = Holidays.read_group(cr, uid, [('employee_id', 'in', ids), ('type', '=', 'remove'), ('date_from', '>=', date_begin.strftime(tools.DEFAULT_SERVER_DATE_FORMAT)), ('date_from', '<=', date_end.strftime(tools.DEFAULT_SERVER_DATE_FORMAT)), ('state', '=', 'validate'), ('payslip_status', '=', False)], ['employee_id'], ['employee_id'], context=context)
```
Comment (atp-odoo) [inline addons/hr_holidays/hr_holidays.py]: better to call read_group only one times with this domain : 
[('employee_id', 'in', ids), ('type', '=', 'remove')]
and then use filtered  on that records with other condition to get approved_leaves_data

File: addons/purchase/partner.py [hunk 4]
```diff
@@ -8,15 +8,16 @@ class res_partner(osv.osv):
...
-                'supplier_invoice_count': Invoice.search_count(cr,uid, [('partner_id', 'child_of', partner_id), ('type','=','in_invoice')], context=context)
-            }
-            for partner_id in ids
-        }
+        res = dict([(id, {'purchase_order_count': 0, 'supplier_invoice_count': 0}) for id in ids])
+        purchase_data = self.pool['purchase.order'].read_group(cr, uid, [('partner_id', 'child_of', ids)], ['partner_id'], ['partner_id'], context=context)
+        for purchase in purchase_data:
+            partner_id = purchase['partner_id'][0]
+            res[partner_id]['purchase_order_count'] = purchase['partner_id_count']
+        invoice_data = self.pool['account.invoice'].read_group(cr, uid, [('partner_id', 'child_of', ids), ('type','=','in_invoice')], ['partner_id'], ['partner_id'], context=context)
```
Comment (atp-odoo) [inline addons/purchase/partner.py]: same as above don't use read_group two times. Use filtered after single read_group.

File: addons/purchase/stock_view.xml [hunk 5]
```diff
@@ -83,5 +83,29 @@
         <menuitem action="action_picking_tree_picking_to_invoice"
             id="menu_action_picking_tree_picking_to_invoice"
             parent="purchase.menu_procurement_management_invoice" sequence="90"/>
+
+        <record id="stock_picking_form_view_inherited" model="ir.ui.view">
+            <field name="name">stock.picking.form.inherit</field>
+            <field name="inherit_id" ref="stock.view_picking_form"/>
+            <field name="model">stock.picking</field>
+            <field name="arch" type="xml">
+                <xpath expr="//sheet/div[@name='button_box']">
```
Comment (atp-odoo) [inline addons/purchase/stock_view.xml]: no position required ?

File: addons/purchase/purchase.py [hunk 6]
```diff
@@ -1757,6 +1757,22 @@ class account_invoice(osv.Model):
...
+    def action_view_purchase_order(self, cr, uid, ids, context=None):
+        ModelData = self.pool['ir.model.data']
+        pur_ids = []
+        result = self.pool['ir.actions.act_window'].for_xml_id(cr, uid, 'purchase', 'purchase_form_action', context=context)
+        for invoice in self.browse(cr, uid, ids, context=context):
+            pur_ids = map(lambda x: x.id, invoice.purchase_ids)
+        if not pur_ids:
+            raise UserError(_('No Purchase Order referenced for this invoice.'))
+        if len(pur_ids) > 1:
+            result['domain'] = [('id', 'in', pur_ids)]
```
Comment (atp-odoo) [inline addons/purchase/purchase.py]: no need to ch eck this condition . always posibility of one purchase order

File: addons/purchase/stock.py [hunk 7]
```diff
@@ -167,6 +167,49 @@ def _get_picking_to_recompute(self, cr, uid, ids, context=None):
...
+    def action_view_purchase_order(self, cr, uid, ids, context=None):
+        pur_ids = set()
+        picking = self.browse(cr, uid, ids)[0]
+        result = self.pool['ir.actions.act_window'].for_xml_id(cr, uid, 'purchase', 'purchase_form_action', context=context)
+        for move in picking.move_lines.filtered(lambda x: x.purchase_line_id):
+            pur_ids.add(move.purchase_line_id.order_id.id)
+        if not pur_ids:
+            raise UserError(_('No Purchase Order Associated with this Incoming Shipment!'))
+        if len(list(pur_ids)) > 1:
+            result['domain'] = [('id', 'in', list(pur_ids))]
```
Comment (atp-odoo) [inline addons/purchase/stock.py]: same here. no need to check this condition . always posibility of one purchase order for incomming shipment.

File: addons/purchase/stock.py [hunk 8]
```diff
@@ -167,6 +167,49 @@ def _get_picking_to_recompute(self, cr, uid, ids, context=None):
                }),
     }
 
+    def action_view_purchase_order(self, cr, uid, ids, context=None):
+        pur_ids = set()
```
Comment (atp-odoo) [inline addons/purchase/stock.py]: why need to take set() here list is not working ?

File: addons/purchase/stock.py [hunk 9]
```diff
@@ -167,6 +167,49 @@ def _get_picking_to_recompute(self, cr, uid, ids, context=None):
...
+                            'left join account_invoice_line ail on (ail.move_id=sm.id) '\
+                            'left join account_invoice ai on (ail.invoice_id = ai.id) '\
+                            'where sp.id in %s and ail.move_id is not null '
+                            'group by sp.id,ai.id', (tuple(ids),))
+            inv_ids =  cr.fetchone()
+            if not inv_ids:
+                raise UserError(_('No invoices created for this Incoming Shipment.'))
+            #choose the view_mode according to invoices
+            if len(inv_ids)>1:
+                result['domain'] = ('id', 'in', inv_ids)
```
Comment (atp-odoo) [inline addons/purchase/stock.py]: same for invoice remove thiscondition.

File: addons/purchase/purchase.py [hunk 10]
```diff
@@ -1757,6 +1757,22 @@ class account_invoice(osv.Model):
...
     }
 
+    def action_view_purchase_order(self, cr, uid, ids, context=None):
+        ModelData = self.pool['ir.model.data']
+        pur_ids = []
+        result = self.pool['ir.actions.act_window'].for_xml_id(cr, uid, 'purchase', 'purchase_form_action', context=context)
+        for invoice in self.browse(cr, uid, ids, context=context):
+            pur_ids = map(lambda x: x.id, invoice.purchase_ids)
+        if not pur_ids:
+            raise UserError(_('No Purchase Order referenced for this invoice.'))
```
Comment (atp-odoo) [inline addons/purchase/purchase.py]: Remove this raise. meaning less to raise Error in stat button.

File: addons/purchase/stock.py [hunk 11]
```diff
@@ -167,6 +167,49 @@ def _get_picking_to_recompute(self, cr, uid, ids, context=None):
...
     }
 
+    def action_view_purchase_order(self, cr, uid, ids, context=None):
+        pur_ids = set()
+        picking = self.browse(cr, uid, ids)[0]
+        result = self.pool['ir.actions.act_window'].for_xml_id(cr, uid, 'purchase', 'purchase_form_action', context=context)
+        for move in picking.move_lines.filtered(lambda x: x.purchase_line_id):
+            pur_ids.add(move.purchase_line_id.order_id.id)
+        if not pur_ids:
+            raise UserError(_('No Purchase Order Associated with this Incoming Shipment!'))
```
Comment (atp-odoo) [inline addons/purchase/stock.py]: same here.Remove this raise. meaning less to raise Error in stat button.

File: addons/purchase/stock.py [hunk 12]
```diff
@@ -167,6 +167,49 @@ def _get_picking_to_recompute(self, cr, uid, ids, context=None):
...
+            inv_ids = []
+            cr.execute('select ai.id as invoice_id from stock_picking sp '\
+                            'left join stock_move sm on (sm.picking_id=sp.id) '\
+                            'left join account_invoice_line ail on (ail.move_id=sm.id) '\
+                            'left join account_invoice ai on (ail.invoice_id = ai.id) '\
+                            'where sp.id in %s and ail.move_id is not null '
+                            'group by sp.id,ai.id', (tuple(ids),))
+            inv_ids =  cr.fetchone()
+            if not inv_ids:
+                raise UserError(_('No invoices created for this Incoming Shipment.'))
```
Comment (atp-odoo) [inline addons/purchase/stock.py]: Remove this raise. meaning less to raise Error in stat button.

File: addons/sale_stock/sale_stock.py [hunk 13]
```diff
@@ -420,6 +420,51 @@ class stock_location_route(osv.osv):
...
+    def action_view_sale_order(self, cr, uid, ids, context=None):
+        sale_ids = set()
+        picking = self.browse(cr, uid, ids)[0]
+        result = self.pool['ir.actions.act_window'].for_xml_id(cr, uid, 'sale', 'action_orders', context=context)
+        if picking.move_lines:
+            for move in picking.move_lines:
+                if move.procurement_id and move.procurement_id.sale_line_id:
+                    sale_ids.add(move.procurement_id.sale_line_id.order_id.id)
+        if not sale_ids:
+            raise UserError(_('No Sale Order Associated with this Delivery Order!'))
```
Comment (atp-odoo) [inline addons/sale_stock/sale_stock.py]: Remove this raise. meaning less to raise Error in stat button.

File: addons/sale_stock/sale_stock.py [hunk 14]
```diff
@@ -420,6 +420,51 @@ class stock_location_route(osv.osv):
...
+        inv_ids = []
+        cr.execute('select ai.id as invoice_id from stock_picking sp '\
+                        'left join stock_move sm on (sm.picking_id=sp.id) '\
+                        'left join account_invoice_line ail on (ail.move_id=sm.id) '\
+                        'left join account_invoice ai on (ail.invoice_id = ai.id) '\
+                        'where sp.id in %s and ail.move_id is not null '
+                        'group by sp.id,ai.id', (tuple(ids),))
+        inv_ids =  cr.fetchone()
+        if not inv_ids:
+            raise UserError(_('No invoices created for this Delivery Order.'))
```
Comment (atp-odoo) [inline addons/sale_stock/sale_stock.py]: Remove this raise. meaning less to raise Error in stat button.

**Comment by spadae22** [comment]
+1

**Pull request closed without merging**